pjmedia/src/pjmedia/txt_stream.c

```diff
@@ -0,0 +1,1054 @@
+ * interval, so we need a larger buffer to handle the burst.
+ */
+#define BUFFER_SIZE             64
+
+/* The number of buffers must be the max redundancy we want to support +
+ * plus one more to store the primary data.
+ */
+#define NUM_BUFFERS             PJMEDIA_TXT_STREAM_MAX_RED_LEVELS + 1
+
+/* 5.4. Compensation for Packets Out of Order:
+ * If analysis of a received packet reveals a gap in the sequence,
+ * we need to wait for the missing packet(s) to arrive.  It is
+ * RECOMMENDED the waiting time be limited to 1 second (1000 ms).
+ */
+#define MAX_RX_WAITING_TIME     1000
+
+/* Maximum number of text packets that can be kept in the jitter buffer.
+ * The value should be roughly our MAX_RX_WAITING_TIME over remote's
+ * buffering time (which unfortunately we don't know).
+ */
+#define JBUF_MAX_COUNT          8
+
+
+/* Buffer to store redundancy and primary data. */
+typedef struct red_buf
+{
+    unsigned timestamp;
+    unsigned length;
+    char     buf[BUFFER_SIZE];
+} red_buf;
+
+typedef struct pjmedia_txt_stream
+{
+    pjmedia_stream_common       base;
+    pjmedia_txt_stream_info     si;             /**< Creation parameter.    */
+
+    pjmedia_clock              *clock;          /**< Clock.                 */
+    unsigned                    buf_time;       /**< Buffering time.        */
+    pj_bool_t                   is_idle;        /**< Is idle?               */
+
+    pj_timestamp                rtcp_last_tx;   /**< Last RTCP tx time.     */
+    pj_timestamp                tx_last_ts;     /**< Timestamp of last tx.  */
+    red_buf                     tx_buf[NUM_BUFFERS];/**< Tx buffer.         */
+    int                         tx_buf_idx;     /**< Index to current buffer*/
+    int                         tx_nred;        /**< Num of redundant data. */
+
+    int                         rx_last_seq;    /**< Sequence of last rx.   */
+    pj_bool_t                   is_waiting;     /**< Is waiting?            */
+    pj_timestamp                rx_wait_ts;     /**< Ts of waiting start.   */
+
+    /* Incoming text callback. */
+    void                      (*cb)(pjmedia_txt_stream *, void *,
+                                    const pjmedia_txt_stream_data *);
+    void                       *cb_user_data;
+
+} pjmedia_txt_stream;
+
+
+static void clock_cb(const pj_timestamp *ts, void *user_data);
+
+
+#include "stream_imp_common.c"
+
+static void on_stream_destroy(void *arg)
+{
+    pjmedia_txt_stream *stream = (pjmedia_txt_stream *)arg;
+
+    if (stream->clock) {
+        pjmedia_clock_destroy(stream->clock);
+        stream->clock = NULL;
+    }
+}
+
+/*
+ * Destroy stream.
+ */
+PJ_DEF(pj_status_t) pjmedia_txt_stream_destroy( pjmedia_txt_stream *stream )
+{
+    pjmedia_stream_common *c_strm = (pjmedia_stream_common *)stream;
+
+    PJ_ASSERT_RETURN(stream != NULL, PJ_EINVAL);
+
+    PJ_LOG(4,(c_strm->port.info.name.ptr, "Stream destroying"));
+
+    if (stream->clock)
+        pjmedia_clock_stop(stream->clock);
```

Comment:
Didn't check the code. Just a heads-up, if there is possibility that the `destroy()` is called from the clock callback, the clock may not be cleaned up properly (may cause pool & handle leaks).

Reply:
Ah yes, I'll put a warning in the doc. As unlikely as it seems that app will want to destroy the stream upon receiving a text, I guess we'll never know what people will do.

Reply:
Some interactive app may hangup the call when receiving a specific text/command?